### PR TITLE
feat(documents): add document search API to stable

### PR DESCRIPTION
**Pull request by tommythorsen**
Only the /documents/search endpoint is available for the time being.

Jira: [SBS-4652](https://cognitedata.atlassian.net/browse/SBS-4652)

### CODE REVIEW

File: packages/stable/src/api/documents/types.ts
Comment (henrist) [inline packages/stable/src/api/documents/types.ts]: Are the types in this file hand-written or somehow generated? If the latter case it would be nice with some instructions for how to keep it up-to-date with future API changes.

I notice that the comments here is not the same as in the API docs, and the values here is no longer correct. It would also be nice if the type names matched the schema used for the API, to reduce cognitive load and reduce type name changes in future automation (as clients of this library will have to migrate their code if they import the types).

File: packages/stable/src/__tests__/api/documentsApi.unit.spec.ts
Comment (henrist) [inline packages/stable/src/__tests__/api/documentsApi.unit.spec.ts]: I think all these ts-ignore rules are needed due to lack of proper types?

If we would need it for these tests, wouldn't a consumer of the library also need it? This makes me think this should be redundant and something we should fix.

I see this in the types, that should be fixed:

```
export type DocumentShapeCoordinates = any[];
```

The API docs behind this is likely the root cause - since it has `type: array` with no further definition

File: packages/stable/src/api/documents/documentsApi.ts
Comment (andersfylling) [inline packages/stable/src/api/documents/documentsApi.ts:36]: You have to add:

`return this.addToMapAndReturn(response.data, response);`

This is so users can get headers or other data. It's built into all the helper methods of the library, but when you write custom methods you need to specify this manually.

File: packages/stable/src/api/documents/documentsApi.ts
Comment (andersfylling) [inline packages/stable/src/api/documents/documentsApi.ts]: I think the constructor can be removed as it doesn't do anything beyond the default behavior.

**Review by andersfylling — CHANGES_REQUESTED** [review]
Few things that have to/can be updated.

File: packages/stable/src/api/documents/types.ts
Comment (henrist) [inline packages/stable/src/api/documents/types.ts]: I also notice this is the sort structure for the playground API, not the v1 API. I think we need to redo this file to get things correct and in sync.

Comment (andersfylling) [reply packages/stable/src/api/documents/types.ts]: we could also write one monster request that touches every field. If the api returns 400, then we know its a little off.

File: packages/stable/src/api/documents/types.ts
Comment (tommythorsen) [reply packages/stable/src/api/documents/types.ts]: It's hand-written. Well, I copied the playground types and modified them to match the v1 spec.

I suppose I could remove these comments. I would not like to manually have to keep them in sync with the spec. That sounds like a lot of work for little benefit.

File: packages/stable/src/api/documents/documentsApi.ts
Comment (tommythorsen) [reply packages/stable/src/api/documents/documentsApi.ts:36]: Oh, maybe I could instead use one of those helper methods that have it built in? I'll look into that

File: packages/stable/src/api/documents/types.ts
Comment (henrist) [reply packages/stable/src/api/documents/types.ts]: I think this is a strong argument for generating the types (#740).

We should focus on the developer experience, so comments should be included. However I'd rather exclude comments than including the wrong comments, if we have to choose.

Since we are not there yet (generating types), can we try to be compatible with it? To preserve developer experience we don't want to rename these later.

Since I spotted at least another defect in the types - maybe @andersfylling can help generating this for a one-shot now? I think this was initially generated for the playground API. That way we get proper names, fix other defects to the types, and will be compatible with future code gen as far as I understand without introducing breaking changes.

File: packages/stable/src/api/documents/documentsApi.ts
Comment (andersfylling) [reply packages/stable/src/api/documents/documentsApi.ts:36]: Maybe you can use `callSearchEndpoint`.

File: packages/stable/src/api/documents/types.ts
Comment (andersfylling) [reply packages/stable/src/api/documents/types.ts]: The v1 spec is funny. I can generate the entire playground api types, without issues, but the v1 might have some weird cyclic reference. Causing a never ending lookup loop (suspicion). It keeps generating forever, ending up with a crash due to memory usage. Even 4GiB isn't enough.

Might be that the script just have to check the type record before exploring a type to avoid redundancy.

Comment (tommythorsen) [reply packages/stable/src/api/documents/types.ts]: The new filter specification is definitely cyclic. In particular, the bool filters refer back to the common filter type

File: packages/stable/src/__tests__/api/documentsApi.unit.spec.ts
Comment (tommythorsen) [reply packages/stable/src/__tests__/api/documentsApi.unit.spec.ts]: The cause for this was simply that the `coordinates` field is optional. I added a `!` behind it, like so:

```js
expect(geoLocation?.coordinates![0]).toEqual(2.324);
```

This removes the need for the `@ts-ignore`

I have not done anything about the `any[]`. Looking at the implementation in the `unstructured-search` repo, it looks like this depends a bit on the type of shape, so it'll get a bit complex if we want to do it exact. I think the `any[]` here is about on par with how it's defined in the contract.

File: packages/stable/src/api/documents/types.ts
Comment (tommythorsen) [reply packages/stable/src/api/documents/types.ts]: I have fixed the sort structure now, and added an integration test for it. There are also integration tests for the other major things.

File: packages/stable/src/api/documents/documentsApi.ts
Comment (tommythorsen) [reply packages/stable/src/api/documents/documentsApi.ts]: Done

File: packages/stable/src/api/documents/documentsApi.ts
Comment (tommythorsen) [reply packages/stable/src/api/documents/documentsApi.ts:36]: I ended up following your first suggestion, which seems to work fine

**Review by andersfylling — APPROVED** [review]
lgtm

**Comment by andersfylling** [comment]
can you add the following to codeowners?

```
/packages/stable/src/api/documents @cognitedata/grep
/packages/stable/src/__tests__/api/documents @cognitedata/grep
```

https://github.com/cognitedata/cognite-sdk-js/pull/750

File: packages/stable/src/__tests__/api/documentsApi.unit.spec.ts
Comment (henrist) [reply packages/stable/src/__tests__/api/documentsApi.unit.spec.ts]: The `any[]` is aligned with the service spec, but I think we should look later to see if we can improve this (in the service spec, and also here). Currently a client have _no idea_ what it might contain (as it is defined as anything) without looking into external resources, while it actually can only hold

```ts
type Position = [number, number]
type Coordinates =
  | Position // Point
  | Position[] // MultiPoint, ListString
  | Position[][] // MultiLineString, Polygon
  | Position[][][] // MultiPolygon
```

This would narrow everything down to numbers and a few lists, and making things a lot simpler to reason about.

File: .github/CODEOWNERS
Comment (henrist) [inline .github/CODEOWNERS]: This does not match the tests files. Maybe it's easily solved by added a `*` at the end of those?

Comment (andersfylling) [reply .github/CODEOWNERS]: Ah, yes, true. It isn't a documents folder, under __tests__, it's a file matching `documents*`

File: packages/stable/src/__tests__/api/documentsApi.int.spec.ts
Comment (henrist) [inline packages/stable/src/__tests__/api/documentsApi.int.spec.ts:6]: What is the canonical source of this test data, in case something gets mixed up (e.g. more files added to the same project that would break these tests) - and for me to actually understand what we are testing against?

File: packages/stable/src/api/documents/types.ts
Comment (henrist) [inline packages/stable/src/api/documents/types.ts]: Is it intentional that this does not have the `Document` prefix as it has in the service spec?

File: packages/stable/src/api/documents/types.ts
Comment (henrist) [inline packages/stable/src/api/documents/types.ts]: This should be optional? `order?:` ?

File: packages/stable/src/api/documents/types.ts
Comment (henrist) [inline packages/stable/src/api/documents/types.ts]: These should not be optional. However the spec is also missing required attribute for these.

File: packages/stable/src/api/documents/types.ts
Comment (henrist) [inline packages/stable/src/api/documents/types.ts]: This is not optional

File: packages/stable/src/api/documents/types.ts
Comment (henrist) [inline packages/stable/src/api/documents/types.ts]: type is not optional

**Review by henrist — CHANGES_REQUESTED** [review]

File: packages/stable/src/api/documents/types.ts
Comment (andersfylling) [reply packages/stable/src/api/documents/types.ts]: Yes, default value is `asc`.

File: .github/CODEOWNERS
Comment (tommythorsen) [reply .github/CODEOWNERS]: I have updated this now

File: packages/stable/src/api/documents/types.ts
Comment (tommythorsen) [reply packages/stable/src/api/documents/types.ts]: fixed

File: packages/stable/src/api/documents/types.ts
Comment (tommythorsen) [reply packages/stable/src/api/documents/types.ts]: fixed

File: packages/stable/src/api/documents/types.ts
Comment (tommythorsen) [reply packages/stable/src/api/documents/types.ts]: fixed

File: packages/stable/src/api/documents/types.ts
Comment (tommythorsen) [reply packages/stable/src/api/documents/types.ts]: I've added the Document prefix now

File: packages/stable/src/__tests__/api/documentsApi.int.spec.ts
Comment (tommythorsen) [reply packages/stable/src/__tests__/api/documentsApi.int.spec.ts:6]: Most of the tests will work regardless of what kinds of files are in the project, as long as there are a few files there. The only exception is maybe `search with query` which requires that there is at least one file with the word "test" somewhere in the project.

**Review by andersfylling — APPROVED** [review]

File: packages/stable/src/__tests__/api/documentsApi.int.spec.ts
Comment (henrist) [reply packages/stable/src/__tests__/api/documentsApi.int.spec.ts:6]: These integration tests have external dependencies and assumtions that is not explicit, so it would help a lot to give a better understanding of this. Have you added new data? Are there seed data available somewhere else? What would we do in case someone wipe the project data? There is a hidden cost here. Might be there are things in the SDK that covers this, but I'm not sure since I haven't touched this here before. I can't find any documentation for integration tests in this repo from a quick look for it.

**Review by henrist — APPROVED** [review]
If types are generated later I'm sure something will be considered breaking from the developer perspective whatever effort we put now. I think the work done so far is very nice!

Currently we are only focusing on happy paths in the tests. With the Documents API we have a few failure cases that's not likely to be that rare. E.g. using a filter that is not supported for a given propertly. I can't understand from this work how that must be handled by a client. Ideally there should be tests for this, and perhaps documentation somewhere.

File: packages/stable/src/api/documents/types.ts
Comment (tommythorsen) [reply packages/stable/src/api/documents/types.ts]: I'm fixing this in the contract first: https://github.com/cognitedata/service-contracts/pull/1614

Comment (tommythorsen) [reply packages/stable/src/api/documents/types.ts]: fixed

**Comment by tommythorsen** [comment]
I think this PR is ready to be merged. @cognitedata/developer-tooling: would one of you be able to review?

**Review by pedroharbs — APPROVED** [review]

**Comment by henrist** [comment]
Since this now include generated code, I advise this is not merged before the code generation logic has been merged. As otherwise it would refer to code and process that does not exist.

**Comment by andersfylling** [comment]
> As otherwise it would refer to code and process that does not exist.

The `gen` part of the filename?

Forgot I had added a gen comment.

But these can be removed. I just forgot about them. Note that if we do get codegen later on and we want the `gen` tag in the filename, removing it now will make it harder to diff later on.

**Comment by andersfylling** [comment]
I'm surprised I didn't see any failing tests in my PR before I merged..

File: packages/stable/src/api/documents/types.gen.ts
Comment (tommythorsen) [inline packages/stable/src/api/documents/types.gen.ts]:
```suggestion
// It has been automatically generated and should not be edited manually
// If you need to make changes to this file, please reach out to tommy.thorsen@cognite.com or anders.fylling@cognite.com
```

**Comment by tommythorsen** [comment]
> Since this now include generated code, I advise this is not merged before the code generation logic has been merged. As otherwise it would refer to code and process that does not exist.

I'm not so fond of keeping PRs open for extended periods of time. Instead, I have modified the comment in the top of the `types.gen.ts` file so that instead of referencing a tool and process that does not exist, it says to reach out to me or Anders.

File: packages/stable/src/api/documents/types.gen.ts
Comment (tommythorsen) [inline packages/stable/src/api/documents/types.gen.ts]:
```suggestion
// If you need to make changes to this file, please reach out to tommy.thorsen@cognite.com
// or anders.fylling@cognite.com
```

**Pull request merged**
